rules/rollup.go

```diff
@@ -0,0 +1,206 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rules
+
+import (
+	"bytes"
+	"errors"
+	"sort"
+	"time"
+
+	"github.com/m3db/m3metrics/filters"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+	"github.com/m3db/m3metrics/policy"
+)
+
+var (
+	emptyRollupTarget rollupTarget
+
+	errNilRollupTargetSchema       = errors.New("nil rollup target schema")
+	errNilRollupRuleSnapshotSchema = errors.New("nil rollup rule snapshot schema")
+	errNilRollupRuleSchema         = errors.New("nil rollup rule schema")
+)
+
+// rollupTarget dictates how to roll up metrics. Metrics associated with a rollup
+// target will be grouped and rolled up across the provided set of tags, named
+// with the provided name, and aggregated and retained under the provided policies.
+type rollupTarget struct {
+	Name     []byte
+	Tags     [][]byte
+	Policies []policy.Policy
+}
+
+func newRollupTarget(target *schema.RollupTarget) (rollupTarget, error) {
+	if target == nil {
+		return emptyRollupTarget, errNilRollupTargetSchema
+	}
+	policies, err := policy.NewPoliciesFromSchema(target.Policies)
+	if err != nil {
+		return emptyRollupTarget, err
+	}
+	tags := make([]string, len(target.Tags))
+	copy(tags, target.Tags)
+	sort.Strings(tags)
+	return rollupTarget{
+		Name:     []byte(target.Name),
+		Tags:     bytesArrayFromStringArray(tags),
+		Policies: policies,
+	}, nil
+}
+
+// sameTransform determines whether two targets have the same transformation.
+func (t *rollupTarget) sameTransform(other rollupTarget) bool {
+	if !bytes.Equal(t.Name, other.Name) {
+		return false
+	}
+	if len(t.Tags) != len(other.Tags) {
+		return false
+	}
+	for i := 0; i < len(t.Tags); i++ {
+		if !bytes.Equal(t.Tags[i], other.Tags[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// clone clones a rollup target.
+func (t *rollupTarget) clone() rollupTarget {
+	policies := make([]policy.Policy, len(t.Policies))
+	copy(policies, t.Policies)
+	return rollupTarget{
+		Name:     t.Name,
+		Tags:     bytesArrayCopy(t.Tags),
+		Policies: policies,
+	}
+}
+
+// rollupRuleSnapshot defines a rule snapshot such that if a metric matches the
+// provided filters, it is rolled up using the provided list of rollup targets.
+type rollupRuleSnapshot struct {
+	name       string
```

Comment:
`name` is actually the name of the rollup rule in this case for ease of use when users change/delete trues (e.g., "rollup by tag a and b"), so it's fine to keep it as a string. The name of the rollup metric is in the rollup target, which is indeed a byte slice.

The only place where a byte slice would make more sense is the namespace name, I can change that sure.